docs/docs/modules/chains/popular/structured_output.mdx

```diff
@@ -34,6 +34,167 @@ import FormatExample from "@examples/chains/openai_functions_structured_format.t
+
+After each LLM response, we'll take our structured data and save it to memory (with a little formatting too). By doing this, we're enabling our LLM to
+remember what the user has said and use that information to generate a response.
+
+The first step is to create our functions. For this we'll use `zod` to create the function schemas.
+
+We're going to create two, one which returns an array of foods with their name, whether or not they're healthy and their color.
+The second simply returns an object containing the final response.
+
+```typescript
+const foodZodSchema = z.object({
+  foods: z
+    .array(
+      z.object({
+        name: z.string().describe("The name of the food item"),
+        healthy: z.boolean().describe("Whether the food is good for you"),
+        color: z.string().optional().describe("The color of the food"),
+      })
+    )
+    .describe("An array of food items mentioned in the text"),
+});
+
+const finalResponseZodSchema = z.object({
+  finalResponse: z.string().describe("The final response"),
+});
+```
+
+Next we'll construct the prompt. We have two "messages", one from the AI with a simple explanation on when to use the functions, and an input variable for the chat history. The second only contains the users question.
+
+```typescript
+const prompt = ChatPromptTemplate.fromMessages([
+  [
+    "ai",
+    `You are a helpful assistant.
+You are provided with two functions: one to use to record what food the user likes, and one to use when replying to user questions.
+Chat history: {history}`,
+  ],
+  ["human", "Question: {input}"],
+]);
+```
+
+Then, define the LLM model and bind the function arguments to it. By using `.bind()` we're adding the functions to the model for every call.
+
+We also must wrap our `zod` schemas with `zodToJsonSchema` to appropriately format them for OpenAI.
+
+```typescript
+const model = new ChatOpenAI({
+  temperature: 0,
+}).bind({
+  functions: [
+    {
+      name: "food_recorder",
+      description: "A function to record what food a user likes",
+      parameters: zodToJsonSchema(foodZodSchema),
+    },
+    {
+      name: "final_response",
+      description: "The function that returns a response to the user",
+      parameters: zodToJsonSchema(finalResponseZodSchema),
+    },
+  ],
+});
+```
+
+Next we can define our memory using the `BufferMemory` class, and the output parser for structured data: `JsonOutputFunctionsParser`.
+
+```typescript
+const outputParser = new JsonOutputFunctionsParser();
+
+const memory = new BufferMemory({
+  memoryKey: "history", // The object key to store the memory under
+  inputKey: "question", // The object key for the input
+  outputKey: "answer", // The object key for the output
+});
+```
+
+Now for the most important step, we'll create a custom function which handles calling the LLM, saving the memory and returning the response.
+
+```typescript
+const executeQuery = async (input: { input: string; history: string }) => {
```

Comment:
This will unfortunately not trace well :(

We would want to do it somewhat like this: https://js.langchain.com/docs/expression_language/cookbook/adding_memory

Also, OpenAI has a special `FunctionMessage` category that might be better?